cmake/MPIConfig.cmake

```diff
@@ -0,0 +1,229 @@
+		endif()
+
+		target_include_directories(${TARGET} PUBLIC ${MPI_${LANGUAGE}_INCLUDE_PATH})
+		
+		if(NOT ${LANGUAGE} STREQUAL CXX)
+			target_compile_definitions(${TARGET} PRIVATE MPI)
+		endif()
+	endmacro()
+
+	# make a MPI version of the thing passed 
+	# also allows switching out sources if needed
+	# INSTALL - causes the new target to get installed in the MPI component to the default location (BINDIR etc)
+	# usage: make_mpi_version(<target> <new name> LANGUAGES <language 1> [<language 2...>] [SWAP_SOURCES <source 1...> TO <replacement source 1...>] INSTALL)
+	function(make_mpi_version TARGET NEW_NAME) 
+	
+		# parse arguments
+		# --------------------------------------------------------------------	
+		cmake_parse_arguments(MAKE_MPI "INSTALL" "" "LANGUAGES;SWAP_SOURCES;TO" ${ARGN})
+	
+		if("${MAKE_MPI_LANGUAGES}" STREQUAL "")
+			message(FATAL_ERROR "Incorrect usage.  At least one LANGUAGE should be provided.")
+		endif()
+		
+		if(NOT "${MAKE_MPI_UNPARSED_ARGUMENTS}" STREQUAL "")
+			message(FATAL_ERROR "Incorrect usage.  Extra arguments provided.")
+		endif()
+	
+		
+		# figure out if it's an object library, and if so, use mpi_object_library()		
+		get_property(TARGET_TYPE TARGET ${TARGET} PROPERTY TYPE)
+		
+		if("${TARGET_TYPE}" STREQUAL "OBJECT_LIBRARY")
+			set(IS_OBJECT_LIBRARY TRUE)
+		else()
+			set(IS_OBJECT_LIBRARY FALSE)
+		endif()
+		
+		if("${ARGN}" STREQUAL "")
+			message(FATAL_ERROR "make_mpi_version(): you must specify at least one LANGUAGE") 
+		endif()
+		
+		# make a new one
+		# --------------------------------------------------------------------
+		if("${MAKE_MPI_SWAP_SOURCES}" STREQUAL "" AND "${MAKE_MPI_TO}" STREQUAL "")
+			copy_target(${TARGET} ${NEW_NAME})
+		else()
+			copy_target(${TARGET} ${NEW_NAME} SWAP_SOURCES ${MAKE_MPI_SWAP_SOURCES} TO ${MAKE_MPI_TO})
+		endif()
+		
+		# this ensures that the MPI version builds after all of the target's dependencies have been satisfied.
+		# Yes it is a bit of an ugly hack, but since we can't copy dependencies, this is the next-best thing.
+		add_dependencies(${NEW_NAME} ${TARGET})
+		
+		# apply MPI flags
+		# --------------------------------------------------------------------
+		foreach(LANG ${MAKE_MPI_LANGUAGES})
+			# validate arguments
+			if(NOT ("${LANG}" STREQUAL "C" OR "${LANG}" STREQUAL "CXX" OR "${LANG}" STREQUAL "Fortran"))
+				message(FATAL_ERROR "make_mpi_version(): invalid argument: ${LANG} is not a LANGUAGE")
+			endif()
+			
+			if(IS_OBJECT_LIBRARY)
+				mpi_object_library(${NEW_NAME} ${LANG})
+			else()
+				string(TOLOWER ${LANG} LANG_LOWERCASE)
+				target_link_libraries(${NEW_NAME} mpi_${LANG_LOWERCASE})
+			endif()
+			
+		endforeach()
+		
+		# install if necessary
+		# --------------------------------------------------------------------
+		if(MAKE_MPI_INSTALL)
+			if("${TARGET_TYPE}" STREQUAL "EXECUTABLE")
+				install(TARGETS ${NEW_NAME} DESTINATION ${BINDIR} COMPONENT MPI)
```

Comment:
I'm a little confused about `COMPONENT MPI`. Could you explain it a bit more?

Reply:
That's used for Amber's install components, it doesn't really matter here, but it isn't doing any harm.  The main use of it is to let you split a program into different components for packaging.